### PR TITLE
fix: openapi schema required field generation

### PR DESCRIPTION
**Pull request by abdulhaq-e**
### Pull Request Checklist

[//]: # "Please review the [Litestar contribution guidelines](https://github.com/litestar-org/litestar/blob/main/CONTRIBUTING.rst) for this repository."

- [x] New code has 100% test coverage
- [x] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [x] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

### Description

The fields defined in a model that uses `pydantic` and `msgspec` DTO factories (and others maybe) are recognized as "not required" by the OpenAPI schema generator. 

This happens because we are relying on a `msgspec` utility function to determine if a field defined in a `Struct` is required. In that function it checks if both `default` and `default_factory` are set to `NODEFAULT` (from `msgspec` internals). 

This leads to wrong behaviour because in the DTO codebase, the `default_factory` is set to `Empty` when no factory is defined. See [here](https://github.com/litestar-org/litestar/blob/92516a55c8cab1e49b2a25e2a7a1e21f223338fb/litestar/contrib/msgspec.py#L36) and [here](https://github.com/litestar-org/litestar/blob/92516a55c8cab1e49b2a25e2a7a1e21f223338fb/litestar/contrib/pydantic.py#L27). The same doesn't apply to `DataclassDTO` in which the `default_factory` is set to `None` but `default` is set to `Empty`.

Overall, I feel the topic of default values in different DTO factories needs an overhaul or clarification (e.g. `default_factory` can be set to `None` per current design, why is it set to `Empty` and why the difference between DTO factories). For now this fixes the OpenAPI schema generator.

[//]: # "Please describe your pull request for new release changelog purposes"
[//]: # "Example: 'This pr adds the capability to use server-sent events, and documents it in the usage docs'"

### Close Issue(s)

[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

### CODE REVIEW

**Review by cofin — APPROVED** [review]
LGTM

File: litestar/_openapi/schema_generation/schema.py [hunk 1]
```diff
@@ -669,12 +670,16 @@ def for_struct_class(self, annotation: type[Struct], dto_for: ForType | None) ->
...
+        def _is_field_requiured(field: FieldInfo) -> bool:
+            return field.required or field.default_factory is Empty
+
         return Schema(
             required=sorted(
                 [
                     field.encode_name
                     for field in msgspec_struct_fields(annotation)
-                    if field.required and not is_optional_union(field.type)
+                    if _is_field_requiured(field=field) and not is_optional_union(field.type)
```
Comment (Goldziher) [inline litestar/_openapi/schema_generation/schema.py:682]: Hmm. Typo

**Review by Goldziher — CHANGES_REQUESTED** [review]

File: litestar/_openapi/schema_generation/schema.py [hunk 2]
```diff
@@ -669,12 +670,16 @@ def for_struct_class(self, annotation: type[Struct], dto_for: ForType | None) ->
         Returns:
             A schema instance.
         """
+
+        def _is_field_requiured(field: FieldInfo) -> bool:
+            return field.required or field.default_factory is Empty
```
Comment (JacobCoffee) [inline litestar/_openapi/schema_generation/schema.py:675]:
```suggestion
        def _is_field_required(field: FieldInfo) -> bool:
            return field.required or field.default_factory is Empty
```

File: litestar/_openapi/schema_generation/schema.py (hunk 1 above)
Comment (JacobCoffee) [inline litestar/_openapi/schema_generation/schema.py:682]:
```suggestion
                    if _is_field_required(field=field) and not is_optional_union(field.type)
```

**Review by JacobCoffee — CHANGES_REQUESTED** [review]

**Review by JacobCoffee — APPROVED** [review]

File: litestar/_openapi/schema_generation/schema.py (hunk 1 above)
Comment (Goldziher) [reply litestar/_openapi/schema_generation/schema.py]: This indicates also that this is not tested. This should have raised.

**Pull request merged**